src/ecdsa/mod.rs

```diff
@@ -292,13 +292,14 @@ impl<C: Signing> Secp256k1<C> {
 
     fn sign_grind_with_check(
         &self,
-        msg: &Message,
+        msg: impl Into<Message>,
         sk: &SecretKey,
         check: impl Fn(&ffi::Signature) -> bool,
     ) -> Signature {
         let mut entropy_p: *const ffi::types::c_void = ptr::null();
         let mut counter: u32 = 0;
         let mut extra_entropy = [0u8; 32];
+        let msg = msg.into();
```

Comment:
I don't get why we have a local var here but on line 262 we just chain the calls?

Reply:
We need the local variable to ensure that the object lives through the entire time that we are using the pointer.

If we're still chaining the calls on line 262 then that also needs to be changed. Thanks for beating me to the review!

Reply:
I'm not sure what you mean by `Into` being "just borrow checker stuff". `Into::into` takes an abstract message by value and returns a `Message` by value. Then `as_c_ptr` borrows this object and returns a raw pointer whose lifetime must not exceed the lifetime of the `Message`.

But if we don't give the `Message` a variable binding, its lifetime will only consist of one line of code. If we give it a binding it'll live until the end of the function.

Reply:
@tcharding note that this is the same underlying issue as this lint: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#temporary-cstring-as-ptr

(one day we can write an attribute to add these lints for user types)

Reply:
Yeah, that'd be wonderful if we could label our own as_ptr methods to trigger a clippy lint.

Reply:
This is just musings and for my education, so please only respond if it amuses you to do so.

When `sign_ecdsa_with_noncedata_pointer` is compiled is the parametr `msg: impl Into<Message>,` just 32 bytes in the stack frame?

> I'm not sure what you mean by Into being "just borrow checker stuff".

I misspoke, FTR I don't know the exact correct terminology for all the parses of the compiler. 

I read the link above but that is different in that a `str` has to have memory backing it but in our case I thought the memory backing it would be the 32 bytes in the stack frame that was passed in as `msg` (in function `sign_ecdsa_with_noncedata_pointer`) - so I can't understand why creating a local variable is fixing the problem.

Said another way, I get that at the end of `msg.into().as_c_ptr()` that there are no guarantees that the pointer is valid, but I don't get why we cannot tell it is valid because we know where the value is on the stack already because it was passed in with the function.

Reply:
(I think I can map C functions to opcodes but I do not know exactly how to map Rust functions to opcodes.)

Reply:
> I don't get why we have a local var here but on line 262 we just chain the calls? Did you mean to change both?

@tcharding it's because here it's used in a loop. `impl Into<Message>` doesn't implement `Copy`, so if we `.into` it inside the loop we lose ownership and cannot proceed more iterations. 

<img width="1386" alt="Screenshot 2024-10-22 at 10 51 27" src="https://github.com/user-attachments/assets/0d0c52d1-db6e-4a7f-9740-97c6cf8d24e1">

By calling `.into` outside of the loop, we get the `Message` type which implements `Copy` so it can be used inside the loop as many times as needed.

---

> We need the local variable to ensure that the object lives through the entire time that we are using the pointer.
>
> If we're still chaining the calls on line 262 then that also needs to be changed. Thanks for beating me to the review!

@apoelstra interesting. Is this due to rustc being able to enforce memory safety guarantees across the ffi boundary?

Reply:
> I read the link above but that is different in that a `str` has to have memory backing it but in our case I thought the memory backing it would be the 32 bytes in the stack frame

How could it be? The 32 bytes in the stack frame belong to an object of a completely different type (an opaque `Into<Message>` vs a `Message`). And we are telling the compiler that we don't need the `Into<Message>` anymore and it can reclaim the memory.

Reply:
I need to read a book on how the Rust compiler works. Thanks for your patience.